packages/core/util/tests/Job-specs.js

```diff
@@ -100,4 +100,28 @@ describe('Job', function () {
 			j.idle(value);
 		});
 	});
+
+	describe('#promise', function () {
+		it('should start job for a resolved promise', function (done) {
+			const j = new Job(() => done());
+			j.promise(Promise.resolve(true));
+		});
+
+		it('should not start job for a rejected promise', function (done) {
+			const j = new Job(() => {
+				done(new Error('Job ran for rejected promise'));
+			});
+			j.promise(Promise.reject(true));
+			setTimeout(done, 10);
+		});
+
+		it('should not start job when stopped before promise resolves', function (done) {
+			const j = new Job(() => {
+				done(new Error('Job ran for stopped promise'));
+			});
+			j.promise(new Promise(resolve => setTimeout(resolve, 20)));
+			setTimeout(() => j.stop(), 10);
+			setTimeout(done, 30);
+		});
+	});
```

Comment:
This test passes but not for the reason you might think.

The `promise()` method doesn't control the execution of the promise but only the execution of the "side effect" passed as the callback to `Job`.  In the test above, the "Previous promise ran" promise still runs (and calls `done()`) but:

* doesn't end the test because `done()` was already called
* doesn't invoke the job because another promise was registered with the job after it

The last test, `it('should not start job when another is started'`, is verifying the "cancelling" feature by passing a new promise to `promise()` thereby ignoring the resolution of the first. If you debug it, you'll notice it still resolves and still calls the `then` but bails when it tests if `this.id === promise`.